### PR TITLE
sql: verify NULLs are not allowed in NOT NULL columns

### LINKED ISSUE #1923: sql: verify NULLs are not allowed in NOT NULL columns

### PR DESCRIPTION
**Pull request by tamird**
Fixes #1923.

### CODE REVIEW

File: sql/insert.go [hunk 1]
```diff
@@ -77,8 +77,17 @@ func (p *planner) Insert(n *parser.Insert) (planNode, error) {
 
 	for rows.Next() {
 		values := rows.Values()
-		if len(values) != len(cols) {
-			return nil, fmt.Errorf("invalid values for columns: %d != %d", len(values), len(cols))
+		for i := 0; i < len(values); i++ {
+			if col := cols[i]; values[i] == parser.DNull && !col.Nullable {
+				return nil, fmt.Errorf("null value in column %q violates not-null constraint", col.Name)
```
Comment (petermattis) [inline sql/insert.go]: Perhaps `s/not-null/NOT NULL/g`?

File: sql/insert.go [hunk 2]
```diff
@@ -77,8 +77,17 @@ func (p *planner) Insert(n *parser.Insert) (planNode, error) {
...
+			}
+		}
+
+		for _, col := range cols[len(values):] {
+			if !col.Nullable {
+				return nil, fmt.Errorf("null value in column %q violates not-null constraint", col.Name)
+			}
+			values = append(values, parser.DNull)
```
Comment (petermattis) [inline sql/insert.go]: Is there another check that all not-NULL columns are being inserted into?

File: sql/insert.go (hunk 1 above)
Comment (tamird) [inline sql/insert.go]: this error string is taken straight from postgres. how strongly do you feel?

**Comment by petermattis** [comment]
LGTM, though I think similar checks are needed in `Update`. Either fix in this PR, or leave #1923 open and fix in a subsequent PR.

File: sql/insert.go (hunk 2 above)
Comment (tamird) [inline sql/insert.go]: what do you mean? this PR may make more sense to review commit-by-commit; this particular change belongs to the first commit.

File: sql/insert.go (hunk 1 above)
Comment (petermattis) [inline sql/insert.go]: I don't feel strongly (there was a question mark there). We've capitalized our SQL keywords elsewhere, though I'm not sure if we've done so in error messages.

File: sql/insert.go (hunk 2 above)
Comment (petermattis) [inline sql/insert.go]: I believe you can do this:

```
  INSERT INTO kv3 VALUES (k) VALUES ('a')
```

We're only inserting into column `k` even though column `v` is NOT NULL.

Comment (tamird) [inline sql/insert.go]: ah, you're right. I'll adjust this.

**Comment by tamird** [comment]
@petermattis PTAL, added the check in update.

File: sql/insert.go [hunk 3]
```diff
@@ -77,8 +77,16 @@ func (p *planner) Insert(n *parser.Insert) (planNode, error) {
...
 		values := rows.Values()
-		if len(values) != len(cols) {
-			return nil, fmt.Errorf("invalid values for columns: %d != %d", len(values), len(cols))
+		for range cols[len(values):] {
+			values = append(values, parser.DNull)
+		}
+
+		for _, col := range tableDesc.Columns {
+			if !col.Nullable {
+				if i, ok := colIDtoRowIndex[col.ID]; !ok || values[i] == parser.DNull {
```
Comment (petermattis) [inline sql/insert.go]: I think it only matters for performance, but we could check that all not-null columns are being inserted into before the table scan. See lines 59-64 where we do something similar for verifying all of the primary-key columns are being inserted into.

**Comment by petermattis** [comment]
LGTM

File: sql/insert.go (hunk 3 above)
Comment (tamird) [inline sql/insert.go]: I'm not sure it's worth the duplication of logic

**Pull request merged**